core/connection.go

```diff
@@ -0,0 +1,119 @@
+package connection
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+/*
+This retry library was inspired by and uses Cenk Alti (https://github.com/cenkalti) backoff library (https://github.com/cenkalti/backoff).
+We would like to thank him for his great work.
+*/
+
+type PermanentError struct {
+	Inner error
+}
+
+func (e PermanentError) Error() string { return e.Inner.Error() }
+func (e PermanentError) Unwrap() error {
+	return e.Inner
+}
+func (e PermanentError) Is(err error) bool {
+	_, ok := err.(PermanentError)
+	return ok
+}
+
+type TransientError struct {
+	Inner error
+}
+
+func (e TransientError) Error() string { return e.Inner.Error() }
+func (e TransientError) Unwrap() error {
+	return e.Inner
+}
+func (e TransientError) Is(err error) bool {
+	_, ok := err.(TransientError)
+	return ok
+}
+
+const MinDelay = 1000
+const RetryFactor = 2
+const NumRetries = 3
+
+// Same as Retry only that the functionToRetry can return a value upon correct execution
+func RetryWithData[T any](functionToRetry func() (T, error), minDelay uint64, factor float64, maxTries uint64) (T, error) {
+	i := 0
+	f := func() (T, error) {
+		var (
+			val T
+			err error
+		)
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
```

Comment:
Do we need this?

Reply:
I added this as a means of catching a panic with one of the contract bindings functions that I ended up fixing so it did not panic (not the one that is in the process of being fixed). @Oppen should we still include a panic catch here?

Reply:
After consulting with @Oppen I've decided to leave it